### ISSUE TITLE
-Wnull-pointer-arithmetic and -Wno-gnu

### ISSUE THREAD

**Issue by nickdesaulniers** [llvm/llvm-project#54444]
The following code w/ ToT LLVM (clang-15) produces the following warnings:

```c
void *x(void) {
    return (char*)0 + 42;
}

void *y(void) {
    return (void*)0 + 42;
}
```
```
<source>:2:21: warning: arithmetic on a null pointer treated as a cast from integer to pointer is a GNU extension [-Wnull-pointer-arithmetic]
    return (char*)0 + 42;
           ~~~~~~~~ ^
<source>:6:21: warning: performing pointer arithmetic on a null pointer has undefined behavior [-Wnull-pointer-arithmetic]
    return (void*)0 + 42;
           ~~~~~~~~ ^
```
for the Linux kernel, we're using `-std=gnu11 -Wno-gnu`, so we should not be observing ANY warnings that we're using GNU C extensions.

I suspect that these two checks should be split into two different named command line flags, such that the one pertaining to GNU C extensions is placed within the `-Wgnu` flag group, such that `-Wno-gnu` implicitly disables it.

cc @AaronBallman @andykaylor 

(link #4440)

As reported in [this thread](https://lore.kernel.org/llvm/20220316183833.1563139-1-arnd@kernel.org/).

**Comment by efriedma-quic** [comment]
See https://reviews.llvm.org/D37042 for why this warning exists in its current form.

**Comment by andykaylor** [comment]
Handling this is beyond my knowledge of the front end diagnostics infrastructure. I had even forgotten that I was the person who added this warning. Based on the discussion in the review, it sounds like marking the warning specifically as a GNU extension will cause it to be an error with -pedantic-errors, but not doing so is the reason it isn't suppressed by -Wno-gnu. Is that right?

@AaronBallman Can you help with this?

**Comment by AaronBallman** [comment]
I think the issue is that the GNU extension diagnostic was added to the `-Wnull-pointer-arithmetic` group, which is a part of `-Wextra` and not `-Wgnu`. I spotted some other diagnostics in the same boat, so I'll try to get this fixed up.

**Comment by nickdesaulniers** [comment]
https://reviews.llvm.org/D122224
